spec/src/test/java/impl/UserQuestionResourceTest.java

```diff
@@ -204,15 +207,81 @@ public void shouldBePossibleToDeleteQuestion() {
     }
+
+    @Test
+    public void shouldYieldCorrectVotesWhenDownVotingAndUpVoting() {
+
+        Auth auth1 = newAuth();
+        Auth auth2 = newAuth();
+        Auth auth3 = newAuth();
+
+        Question question = questionResource.createQuestion(newAuth(), getQuestion("my question title", "my question")).toBlocking().singleOrDefault(null);
+
+        voteAndAssertSuccess(userQuestionResource::downVoteQuestion, auth1, question, -1);
+        voteAndAssertFailure(userQuestionResource::downVoteQuestion, auth1, question);
+        voteAndAssertSuccess(userQuestionResource::upVoteQuestion, auth1, question, 0);
+        assertNoVote(auth1, question);
+        voteAndAssertSuccess(userQuestionResource::upVoteQuestion, auth1, question, 1);
+        voteAndAssertFailure(userQuestionResource::upVoteQuestion, auth1, question);
+
+        voteAndAssertSuccess(userQuestionResource::upVoteQuestion, auth2, question, 2);
+
+        voteAndAssertSuccess(userQuestionResource::upVoteQuestion, auth3, question, 3);
+
+        voteAndAssertSuccess(userQuestionResource::downVoteQuestion, auth2, question, 2);
+        assertNoVote(auth2, question);
+
+        voteAndAssertSuccess(userQuestionResource::downVoteQuestion, auth3, question, 1);
+        assertNoVote(auth2, question);
+    }
+
+    @Test
+    public void shouldThrowErrorWhenUserVotesOnOwnAnswer() {
+
+        Auth     auth     = newAuth();
+        Question question = questionResource.createQuestion(auth, getQuestion("my question title", "my question")).toBlocking().singleOrDefault(null);
+
+        voteAndAssertFailure(userQuestionResource::downVoteQuestion, auth, question);
+        voteAndAssertFailure(userQuestionResource::upVoteQuestion, auth, question);
+    }
+
+    private void voteAndAssertSuccess(BiFunction<Auth, Long, Observable<Void>> call, Auth auth, Question question, int expectedVotes) {
+        call.apply(auth, question.getId()).test()
+            .awaitTerminalEvent()
+            .assertNoErrors();
+        assertThat(questionResource.getQuestionById(question.getId()).test().awaitTerminalEvent().getOnNextEvents())
+            .extracting(Question::getVotes)
+            .containsExactly(expectedVotes);
+    }
+
+    private void voteAndAssertFailure(BiFunction<Auth, Long, Observable<Void>> call, Auth auth, Question question) {
+        assertThatExceptionOfType(WebException.class)
+            .isThrownBy(() -> call.apply(auth, question.getId()).toBlocking().single())
+            .satisfies(e -> {
+                assertThat(e).isInstanceOf(WebException.class);
+                assertThat(e).hasFieldOrPropertyWithValue("status", BAD_REQUEST);
+                assertThat(e).hasFieldOrPropertyWithValue("error", INVALID_VOTE);
+            });
+    }
+
+    private void assertNoVote(Auth auth, Question answer) {
+        questionVoteDao.findVote(auth.getUserId(), answer.getId()).test().awaitTerminalEvent()
+            .assertNoValues();
+    }
+
+    private Auth newAuth() {
```

Comment:
I think we in a later pr should create reusable code :)

Reply:
Yup. Extracting stuff to some test util would be nice.